### DISCUSSION TITLE
Why all users are logged out during minor upgrades?

### DISCUSSION THREAD

**Discussion by haininghu** [keycloak/keycloak#21205] [Q&A] [answered]
Hello everyone,

in order to prepare an upgrade of our running Keycloak instance (18.x to 21.x) I started reading the upgrading guides. Right at the beginning, I stumbled across the following https://www.keycloak.org/docs/latest/upgrading/index.html#_prep_migration:

> In a minor upgrade of Keycloak, all user sessions are lost. After the upgrade, all users will have to log in again.

My question is: Is there really no way to keep the active user sessions during an upgrade?

Thanks in Advance

Henning

**Comment by sschu** [comment, marked as answer]
I see two main options:
a) If the update does not contain an update of the embedded Infinispan version, old and new Keycloak node can probably just take over the data and sessions are kept. This is not guaranteed to work so you should just test this. It most likely won't work on major updates that include Infinispan updates.
b) store all sessions in an external Infinispan. The downside is you have to operate it.

**Reply by haininghu** [reply]
Thanks for your reply. So the only save solution would be an external Infinispan. And as long as we don't mess up that external Infinispan, we won't lose our sessions while updating Keycloak, right?

**Reply by sschu** [reply]
Yes. But you still might have to update your Infinispan as well which might cause session loss...

**Reply by adnanvrbic** [reply]
@sschu but offlineSessions should be loaded from the database on the clean start, right? So even if you go down (keycloak and external infinispan) on the new clean start, Keycloak will load sessions from the database, on the older version it will do on keycloak start, on newer versions it will do a lazy load. Either way, the upgrade in most cases should not affect offline sessions, at least that was not the case for us.

**Reply by sschu** [reply]
@adnanvrbic Correct, I was only talking about normal sessions.

**Reply by adnanvrbic** [reply]
Oh sorry I missed that part, I thought that we are talking about offlineSessions.

**Reply by darius-m** [reply]
You can also add persistency for your (external) Infinispan instances, which should allow loading the cache data from a store (database / file) if the Infinispan service restarts.

**Reply by xgp** [reply]
@sschu Thanks for your answer here. Can you elaborate on **why** an embedded Infinispan upgrade doesn't work? Is there some fundamental incompatibility that is detected (like a version or magic number) that causes them not to work together? 

I have experienced the same thing when upgrading a cluster, and I notice that Infinispan in the new Keycloak will fail with one of many exceptions. However, I assumed, since they use the same protocol as communicating with an external Infinispan, that there should be no difference between the two.

Also, do you have an understanding of **why** using an external Infinispan allows no-downtime upgrades to work, even when Keycloak and the external Infinispan are using different versions?

I assume both (Infinispan embedded in Keycloak, Keycloak connected to external Infinispan) use HotRod to communicate, which appears to be a stable protocol. I just can't figure out what is different.

**Reply by sschu** [reply]
To my knowledge it is exactly as you suspected, there are some magic numbers in message headers. And the protocol used is different, an embedded Infinispan uses JGroups while talking to an external Infinispan would use HotRod. Therefore, you can update Keycloak without downtime if you don't touch the external Infinispan. However, when you update that one you will have a downtime I suspect (since Infinispan internally uses JGroups) so in a way you are just shifting the problem.

**Reply by xgp** [reply]
Thanks @sschu 

I was reading this document about using the operator to do [Infinispan cluster upgrades](https://infinispan.org/docs/infinispan-operator/main/operator.html#cluster-upgrades_upgrading-clusters) and it indicates:

> HotRodRolling
> Upgrades Infinispan clusters without service downtime.

It apparently does this by creating a new, remote cluster (based on our assumptions, that would be using HotRod), transfers the cache from the existing cluster to the new cluster, and then redirects the clients to the new cluster. This seems to validate the JGroups versus HotRod assumption. However, would there be any way to build the logic of the operator into Keycloak itself?

**Reply by sschu** [reply]
You mean for the embedded Infinispan? I don't think this would work since HotRod is not even used. With an external Infinispan trying out this rolling update would be an interesting experiment.

**Reply by xgp** [reply]
Yeah. I guess I had some fantasy about telling Keycloak that it's a new version, and then exposing itself via HotRod to the "old" cluster and forcing the cache to sync. Essentially doing the same thing that the Infinispan operator does. However, probably a lot of work that few people would use.

I'll play around with the Infinispan operator to see if the rolling update works when Keycloak is connected.

**Reply by xgp** [reply]
Somewhat of a detour, but I did a quick PoC that essentially eliminates Infinispan clustering, and thus upgrade problems, by converting all keycloak "distributed" caches to local ones with a JDBC store. That store gets used by all nodes, even though they don't know about each other. So far, there appears to be no collision/concurrency problems.

https://github.com/xgp/keycloak-experiments/tree/main/cache-persistence

I also ran it on a `20.x` to `22.x` upgrade with multiple nodes, and there were no problems.

**Reply by sschu** [reply]
Interesting idea. Didn't you have problems with db migrations in KC21? Also I am wondering how Infinispan actually notices changes in the db when reading. Is it essentially a read-through cache? ;)

**Reply by xgp** [reply]
Migrations are not related. That experiment creates `ispn_*` tables in the db and uses them to store the cache entries.

**Reply by sschu** [reply]
Well, migrations might also prevent you from doing an update without downtime in addition to the Infinispan issue.
Furthermore, for me it is still unclear how you make sure you read the current data when reading from the cache. Suppose:

1. User goes to node A and creates a session.
2. User comes to node B and updates the session (loading the state before the update from the db).
3. User comes to node A again - how does A notice the changes from B?

Unless every nodes always reads all data from the database (essentially not caching at all), node A might miss changes done by node B.

**Reply by xgp** [reply]
> Unless every nodes always reads all data from the database (essentially not caching at all)

With `passivation=false` and `memory size=0`, the writes/reads should always go to/from the db.

To be clear, it wouldn't be the most performant solution, but it solves the problems that most of our customers have with Keycloak (losing sessions on restart and upgrade, operational complexity) in exchange for higher db load.

**Reply by sschu** [reply]
That makes sense, thanks.
